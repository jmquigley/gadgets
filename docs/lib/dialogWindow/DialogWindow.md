<a name="module_DialogWindow"></a>

## DialogWindow
A modal dialog window.  This component uses the [react-modal](https://github.com/reactjs/react-modal)library.  The component presents a propup window with a title bar and aclose button.## Screen:<img src="https://github.com/jmquigley/gadgets/blob/master/images/dialogWindow.png" width="50%" />## Examples:```javascriptimport {DialogWindow} from 'gadgets';<DialogWindow   height="600px"   icon="plane"   onClose={this.handleCloseDialog}   onOpen={this.handleOpenDialog}   show={true}   title="Demo Dialog Window"   width="600px">    <span>Dialog Content</span></DialogWindow>```## API#### Events- `onClose()` - callback invoked when the dialog window is closed.- `onOpen()` - callback invoked when the dialog window is opened.#### Styles- `ui-dialogwindow` - Placed on the main `<div>` component that surroundsthe whole component.- `ui-dialogwindow-content` - Placed on the `<div>` that surrounds the childcontent given to the window.#### Properties- `height {string} ('400px')` - the height, in pixels, of the dialog area- `icon {string} ('window-restore')` - A font awesome icon that will be onthe right side of the title bar- `show {boolean} (false)` - when set to true the window is shown, otherwiseit is hidden.- `title {string} ('Dialog Window')` - A text string shown within the titlebar of the dialog window- `width {string} ('400px')` - the width, in pixels, of the dialog area

