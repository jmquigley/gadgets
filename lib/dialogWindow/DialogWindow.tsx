/**
 * {description}
 *
 * #### Examples:
 *
 * ```javascript
 * import {Button} from 'gadgets';
 * <Button iconName="cab" onClick={someFunction} />
 * ```
 *
 * #### Events
 * - `{name}` - {description}
 *
 * #### Styles
 * - `` - {description}
 *
 * #### Properties
 * - `{name}: {datatype}` - {description}
 *
 * @module DialogWindow
 */

'use strict';

const ReactModal = require('react-modal');

import autobind from 'autobind-decorator';
import {cloneDeep} from 'lodash';
import * as React from 'react';
import {nilEvent} from 'util.toolbox';
import {
	BaseComponent,
	BaseProps,
	getDefaultBaseProps,
	Wrapper
} from '../shared';

export interface DialogWindowProps extends BaseProps {
	icon?: string;
	onClose?: any;
	onOpen?: any;
	show?: boolean;
	title?: string;
}

export interface DialogWindowState {
	showModal?: boolean;
}

export function getDefaultDialogWindowProps(): DialogWindowProps {
	return cloneDeep(Object.assign({},
		getDefaultBaseProps(), {
			icon: 'bomb',
			obj: 'DialogWindow',
			onClose: nilEvent,
			onOpen: nilEvent,
			show: false,
			title: 'Dialog Window'
		})
	);
}

export class DialogWindow extends BaseComponent<DialogWindowProps, DialogWindowState> {

	public static defaultProps: DialogWindowProps = getDefaultDialogWindowProps();

	private _customStyle: any = {
		content: {
			top: '40%',
			left: '50%',
			right: 'auto',
			bottom: 'auto',
			marginRight: '-50%',
			transform: 'translate(-50%, -50%)',
			padding: '15px'
		}
	};

	constructor(props: DialogWindowProps) {
		super(props, DialogWindow.defaultProps.style);

		this._classes.add('ui-dialogwindow');

		this.state = {
			showModal: this.props.show
		};

		this.componentWillUpdate(this.props);
	}

	@autobind
	private handleClose() {
		this.props.onClose();
	}

	@autobind
	private handleOpen() {
		this.props.onOpen();
	}

	public componentWillReceiveProps(nextProps: DialogWindowProps) {
		if (this.props.show !== nextProps.show) {
			this.setState({showModal: nextProps.show});
		}
	}

	public render() {
		return(
			<Wrapper {...this.props} >
				<ReactModal
					contentLabel={this.props.title}
					isOpen={this.state.showModal}
					onAfterOpen={this.handleOpen}
					onRequestClose={this.handleClose}
					shouldCloseOnOverlayClick={false}
					style={this._customStyle}
				>
					{this.props.children}
				</ReactModal>
			</Wrapper>
		);
	}
}
