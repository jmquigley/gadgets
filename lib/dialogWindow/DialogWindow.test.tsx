"use strict";

import {shallow} from "enzyme";
import assert from "power-assert";
import * as React from "react";
import DialogWindow from "./DialogWindow";

test("Test creation of a default DialogWindow control", () => {
	const ctl = shallow(
		<DialogWindow
			className='test-class'
			height='600px'
			icon='plane'
			show={true}
			title='Demo Dialog Window'
			width='600px'
		>
			<span>Dialog Content</span>
		</DialogWindow>
	);

	assert(ctl);
	expect(ctl).toMatchSnapshot();
});
