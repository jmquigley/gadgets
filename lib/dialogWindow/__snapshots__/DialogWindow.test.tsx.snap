// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test creation of a default DialogWindow control 1`] = `
<Wrapper
  backgroundColor={null}
  border={null}
  borderColor={null}
  bottom={null}
  className="test-class"
  color={null}
  controlled={true}
  direction="right"
  disabled={false}
  err={null}
  errorMessage={null}
  focus={false}
  height="600px"
  icon="plane"
  id={null}
  left={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noborder={false}
  noedit={false}
  nohover={false}
  notheme={false}
  notooltip={false}
  obj="DialogWindow"
  onClose={[Function]}
  onError={[Function]}
  onOpen={[Function]}
  padding={null}
  reset={false}
  ripple={false}
  selected={false}
  show={true}
  sizing="normal"
  style={Object {}}
  testing={true}
  theme={null}
  title="Demo Dialog Window"
  tooltip={null}
  top={null}
  visible={true}
  width="600px"
  xcss={null}
>
  <Modal
    ariaHideApp={false}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel="Demo Dialog Window"
    isOpen={true}
    onAfterOpen={[Function]}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={false}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "borderRadius": "0",
          "bottom": "auto",
          "left": "50%",
          "padding": "0",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        },
      }
    }
  >
    <DialogWindow__DialogWindowView
      backgroundColor={null}
      border={null}
      borderColor={null}
      bottom={null}
      className="ui-dialogwindow test-class"
      color={null}
      controlled={true}
      direction="right"
      disabled={false}
      err={null}
      errorMessage={null}
      focus={false}
      height="600px"
      icon="plane"
      id={null}
      left={null}
      location="none"
      maxHeight={null}
      maxWidth={null}
      minHeight={null}
      minWidth={null}
      noborder={false}
      noedit={false}
      nohover={false}
      notheme={false}
      notooltip={false}
      obj="DialogWindow"
      onClose={[Function]}
      onOpen={[Function]}
      padding={null}
      ripple={false}
      selected={false}
      show={true}
      sizing="normal"
      style={Object {}}
      testing={true}
      theme={null}
      title="Demo Dialog Window"
      tooltip={null}
      top={null}
      visible={true}
      width="600px"
      xcss={null}
    >
      <DialogWindow__ItemView
        backgroundColor={null}
        border={null}
        borderColor={null}
        bottom={null}
        className="test-class"
        color={null}
        controlled={true}
        direction="right"
        disabled={false}
        err={null}
        errorMessage={null}
        focus={false}
        height="600px"
        icon="plane"
        id={null}
        left={null}
        leftButton={
          <Icon
            backgroundColor={null}
            border={null}
            borderColor={null}
            bottom={null}
            className={null}
            color={null}
            controlled={true}
            direction="right"
            disabled={false}
            err={null}
            errorMessage={null}
            focus={false}
            height={null}
            iconName="plane"
            id={null}
            imageFile=""
            left={null}
            location="none"
            maxHeight={null}
            maxWidth={null}
            minHeight={null}
            minWidth={null}
            noborder={false}
            noedit={false}
            nohover={false}
            notheme={false}
            notooltip={false}
            obj="Icon"
            padding={null}
            ripple={false}
            selected={false}
            sizing="normal"
            style={Object {}}
            testing={true}
            theme={null}
            tooltip={null}
            top={null}
            visible={true}
            width={null}
            xcss={null}
          />
        }
        location="none"
        maxHeight={null}
        maxWidth={null}
        minHeight={null}
        minWidth={null}
        noborder={false}
        noedit={false}
        nohover={true}
        noripple={true}
        notheme={false}
        notooltip={false}
        obj="DialogWindow"
        onClose={[Function]}
        onOpen={[Function]}
        padding={null}
        rightButton={
          <ForwardRef
            iconName="times"
            onClick={[Function]}
          />
        }
        ripple={false}
        selected={false}
        show={true}
        sizing="normal"
        style={Object {}}
        testing={true}
        theme={null}
        title="Demo Dialog Window"
        tooltip={null}
        top={null}
        visible={true}
        width="600px"
        xcss={null}
      >
        <span>
          Dialog Content
        </span>
      </DialogWindow__ItemView>
      <DialogWindow__DialogWindowContent
        className="ui-dialogwindow-content"
      >
        <span>
          Dialog Content
        </span>
      </DialogWindow__DialogWindowContent>
    </DialogWindow__DialogWindowView>
  </Modal>
</Wrapper>
`;

exports[`Test retrieval of DialogWindow props object 1`] = `
Object {
  "backgroundColor": null,
  "border": null,
  "borderColor": null,
  "bottom": null,
  "children": null,
  "className": null,
  "color": null,
  "controlled": true,
  "direction": "right",
  "disabled": false,
  "err": null,
  "errorMessage": null,
  "focus": false,
  "height": "400px",
  "icon": "window-restore",
  "id": null,
  "left": null,
  "location": "none",
  "maxHeight": null,
  "maxWidth": null,
  "minHeight": null,
  "minWidth": null,
  "noborder": false,
  "noedit": false,
  "nohover": false,
  "notheme": false,
  "notooltip": false,
  "obj": "DialogWindow",
  "onClose": [Function],
  "onOpen": [Function],
  "padding": null,
  "ripple": false,
  "selected": false,
  "show": false,
  "sizing": "normal",
  "style": Object {},
  "testing": true,
  "theme": null,
  "title": "Dialog Window",
  "tooltip": null,
  "top": null,
  "visible": true,
  "width": "400px",
  "xcss": null,
}
`;

exports[`Test retrieval of DialogWindow props object 2`] = `
Object {
  "children": null,
  "icon": null,
  "showModal": false,
  "sizing": "normal",
  "style": Object {},
}
`;
